### PR TITLE
Gather test failing

### PR DESCRIPTION
**Pull request by InnocentBug**
Here is what the problem is.

I reduced it to a very simple Graph, 2 nodes, and they are interconnected.
And the atoms, I am gather from is just index 1, but I gather to all nodes of the graph.

I use the sequential feature to really see what is going on.

With 4 channels, gathering and gathering from index.
I get a ptensolayer with a single atom of value [0, 1, 2, 3]

To calculate the gradient. I sum the ptensorlayer up and calculate the gradient.

The result before summing is `[[0, 0, 0,0], [0,1,2,3]]` as expected and the sum is 6.

but the gradient is numerically on all inputs.
But I think it should be 1, and that is also what I get from my manual calculation.

From here on, I mark in the code, what I think are the key aspects.

Tested on CPU only for now.

### CODE REVIEW

**Comment by InnocentBug** [comment]
In the backward pass of `ptensolayer0` for gather, I get `g` and `g_view` to be a tensor of [[1, 1, 1, 1], [1,1,1,1]].
Which I guess makes sense, we just have to gather it back up, so that only the contributions from index 1 matter.

File: layers/Ptensors0.hpp [hunk 1]
```diff
@@ -330,8 +330,9 @@ namespace ptens{
     template<typename OUTPUT>
     void add_gather_back(const OUTPUT& x, const LayerMap& map){
       auto plan=GatherPlanFactory::gather_map0(map,x.atoms,atoms,x.getk(),0);
-      if constexpr(std::is_same<OUTPUT,Ptensors0<TYPE> >::value)
+      if constexpr(std::is_same<OUTPUT,Ptensors0<TYPE> >::value){
```
Comment (InnocentBug) [inline layers/Ptensors0.hpp:333]: This is where the work actually happens on the C++ side.
I am not quite sure the gather plan works, that is what confuses me.

Here is what I found out so far.

When we enter this member function, the object is numerically but of the right dimension for the output.
And x is as expected the `g_view` value that we put in.

The in plan.out() is: `[ 0 1 1 1 0 ]` and the plan.in() is `[ 0 1 0 1 0 ]`.

I checked the value of the result of `x.reduce0(plan.out()`. and is the same numerical 0 as before `[ 1.70958e-43 1.96462e-42 1.13505e-43 0 ]`.

And after the broadcast0 call the this object is still numerically zero.

**Comment by InnocentBug** [comment]
The first thing, that comes to mind of what may go wrong is that I am using the wrong `atoms` when constructing `g_view` in the backwards call.
Can you double check that please?

**Comment by InnocentBug** [comment]
The other thing, that comes to mind is that somewhere in the `view` calling along the way, we are writing into a copied object instead of the one we want to point a reference to.
The one that comes to mind would be from Ptensor0.hpp:261
```
    void broadcast0(const BASE& x, const vector<int>& ix, const int offs){
      assert(ix.size()==1);
      view(offs,x.dim(0))+=x.view1();
    }
```
If the `+=` is on a copy and not a reference that may be the problem.

**Comment by InnocentBug** [comment]
What got me generally a little confused here is that we have 2 headers one is `tensor/Ptensor0.hpp` and the other one is `layers/Ptensors0.hpp` (with and s).
And we kind of use both, but that has me a bit puzzled.

**Comment by InnocentBug** [comment]
Any insight for me to understand this more.
And/or idea of what going wrong here?

File: python/pytest/test_tests.py [hunk 2]
```diff
@@ -29,36 +29,53 @@ def test_bug1(device):
...
-#         loss.backward()
-#         xgrad=x.grad
-
-
-#         fn = lambda x: cls.gather(atoms2, x)
-#         xgrad2 = numerical_grad_sum(fn, x, self.h)
-        
-#         assert torch.allclose(xgrad, xgrad2, rtol=1e-2, atol=1e-2)
+graph_atoms_list= [
+    (ptens.ggraph.from_edge_index( torch.Tensor([[0, 1], [1, 0]]).int() ),
```
Comment (InnocentBug) [inline python/pytest/test_tests.py:33]: @risi-kondor here is where I create the Graph and Atomspack that form the ptensorlayer0 for the test.

In the previous test both were random.
But I realized that this one already creates the problem.

**Comment by InnocentBug** [comment]
I think, I am realizing why `gradcheck` fails, but my manual test passes.

I think gradcheck expects the variable we are differentiating for to be the first argument.
But that is atomspack, and not the layer.

But let me check that more.

**Comment by InnocentBug** [comment]
Ok, I figured this one kind of out.

@risi-kondor you were right something within `torch.sum` makes this go wrong.
When I use the `L1Loss` with `sum` reduction and a 0 reference it is the same as `torch.sum` but it works.
So, something within this doesn't work.

I also realize that something goes wrong when we use `ptensorlayer2`.

I checked it in with the work-around for now.

Let's merge this one as "green" in, since the gather function seems to work.
And open new ones for the issue with `torch.sum` and `ptensorlayer2`.
